ic-agent/src/agent/mod.rs

```diff
@@ -1020,6 +1126,72 @@ pub fn signed_request_status_inspect(
     Ok(())
 }
 
+#[derive(Clone)]
+struct SubnetCache {
+    subnets: TimedCache<Principal, Arc<Subnet>>,
+    canister_index: RangeInclusiveMap<Principal, Principal, PrincipalStep>,
+}
+
+impl SubnetCache {
+    fn new() -> Self {
+        Self {
+            subnets: TimedCache::with_lifespan(300),
+            canister_index: RangeInclusiveMap::new_with_step_fns(),
+        }
+    }
+
+    fn get_subnet_by_canister(&mut self, canister: &Principal) -> Option<Arc<Subnet>> {
+        self.canister_index
+            .get(canister)
+            .and_then(|subnet_id| self.subnets.cache_get(subnet_id).cloned())
+            .filter(|subnet| subnet.canister_ranges.contains(canister))
+    }
+
+    fn insert_subnet(&mut self, subnet_id: Principal, subnet: Arc<Subnet>) {
+        self.subnets.cache_set(subnet_id, subnet.clone());
+        for range in subnet.canister_ranges.iter() {
+            self.canister_index.insert(range.clone(), subnet_id);
+        }
```

Comment:
In the case of canister migration(where a canister is moved from subnet A to B) and subnet splitting(where subnet A is split into subnet A’ and B), the subnet’s canister ranges can change.

Suppose the agent talks to canister_1 and canister_2 in the subnet A. Then canister_2 is moved to B. If the agent keeping making queries to canister_1 then the cache for subnet A will always be there when the agent makes queries to canister_2. Thus a new remote request for fetching node keys for canister_2 will not be triggered and in turn fail all queries to canister_2.

I am not sure how RangeInclusiveMap exactly works. If stale ranges for A can automatically be removed and then new ones for A are inserted when we refetch node keys for A, this will solve the problem that canister_2 is always mapped to subnet A. If the agent does not make queries will canister_1 anymore, it is true that the canister_index is stale. However, it won't be a problem for canister_2, because the cache miss will trigger remote fetch of subnet as it will work as long as overlapped ranges between A and B can be overridden.

Do you think this solution is feasible? In terms of code, what do you think of first removing all existing ranges mapped to the subnet_id and then do the for loop to insert new ones?

cc @ericswanson-dfinity @adamspofford-dfinity @roelstorms

Reply:
@krpeacock I see you use a direct mapping from canister to subnet in agent-js for the cache. 
In the case of canister ranges changes,  the mapping may be invalid. Updating the direct mapping from the canister to the new subnet should be easy, right?

Reply:
@ShuoWangNSL When a shrunken range for A is reinserted, it will indeed not update the rangemap index to account for canister_2 not being in it. However, there's a final check that canister_2 is in A's range, and if this check fails, it is still a cache miss rather than an error, which will trigger looking up B and updating the index.

Reply:
It could be written to remove old indexes, but since there's going to be range-related cache misses *anyway*, I'd rather they result in more hash lookups than more network requests.

Reply:
> there's a final check that canister_2 is in A's declared range, and when this check fails, it is still a cache miss rather than an error, which will trigger looking up B and updating the index.

I got it. If the cache of subnet A expires, not only the node keys but also the subnet ranges will be updated by a new read_state call. Then although the canister_index directs canister_2 to the cache of subnet A but the final check of canister ranges will fail. 

Thanks for your explanation. It makes sense to me.